File: arithmetic_evaluator.py
# public class Number {
#   public int value;
#   public Number(int value) {
#     this.value = value;
#   }
# }
class Number:
    def __init__(self, value):
        self.value = value
    def toNumber(self):
        return self.value

    
class Plus:
    def __init__(self, left, right):
        self.left = left
        self.right = right
    def toNumber(self):
        return self.left.toNumber() + self.right.toNumber()
    
class Multiply:
    def __init__(self, left, right):
        self.left = left
        self.right = right
    def toNumber(self):
        return self.left.toNumber() * self.right.toNumber()

class Subtract:
    def __init__(self, left, right):
        self.left = left
        self.right = right
    def toNumber(self):
        return self.left.toNumber() - self.right.toNumber()

class UnaryMinus:
    def __init__(self, child):
        self.child = child
    def toNumber(self):
        return -self.child.toNumber()
    
# 8
# Number(8)

# 2 + 3
# Plus(Number(2), Number(3))

# 3 * 4 + 5
# Plus(Multiply(Number(3), Number(4)),
#      Number(5))

# 4 - 5
# Subtract(Number(4), Number(5))

# -(2 + 2)
# UnaryMinus(Plus(Number(2),
#                 Number(2)))

# takes an AST, returns a Python number
def evaluate(node):
    if isinstance(node, Number):
        return node.value
    elif isinstance(node, Plus):
        leftResult = evaluate(node.left)
        rightResult = evaluate(node.right)
        return leftResult + rightResult
    elif isinstance(node, Multiply):
        leftResult = evaluate(node.left)
        rightResult = evaluate(node.right)
        return leftResult * rightResult
    elif isinstance(node, Subtract):
        return evaluate(node.left) - evaluate(node.right)
    elif isinstance(node, UnaryMinus):
        return -evaluate(node.child)

File: arithmetic_evaluator.pl
% exp ::= number(int) |
%         plus(exp, exp) |
%         multiply(exp, exp) |
%         subtract(exp, exp) |
%         unaryMinus(exp)

% binop ::= plus | multiply | subtract
% exp :: = number(int) | binaryOperation(exp, binop, exp) | unaryMinus(exp)

% evaluateTwo(InExp1, InExp2, E1Result, E2Result, Operation, Result)
evaluateTwo(E1, E2, E1Result, E2Result, Operation, Result) :-
    evaluate(E1, E1Result),
    evaluate(E2, E2Result),
    Result is Operation.
    
% evaluate(InputExp, OutputInt)
evaluate(number(Int), Int).
evaluate(plus(E1, E2), Result) :-
    evaluateTwo(E1, E2, E1Result, E2Result, +(E1Result, E2Result), Result).
    %% evaluate(E1, E1Result), % E1Result = evaluate(E1) % python
    %% evaluate(E2, E2Result), % E2Result = evaluate(E2) % python
    %% is(Result, +(E1Result, E2Result)).
    %% % Result is E1Result + E2Result.
evaluate(multiply(E1, E2), Result) :-
    evaluate(E1, E1Result), % E1Result = evaluate(E1) % python
    evaluate(E2, E2Result), % E2Result = evaluate(E2) % python
    Result is E1Result * E2Result.
evaluate(subtract(E1, E2), Result) :-
    evaluate(E1, E1Result), % E1Result = evaluate(E1) % python
    evaluate(E2, E2Result), % E2Result = evaluate(E2) % python
    Result is E1Result - E2Result.
evaluate(unaryMinus(E), Result) :-
    evaluate(E, EResult),
    Result is -EResult.
